backend/src/composite/mod.rs

```diff
@@ -231,6 +244,120 @@ fn set_size<F: Clone>(pil: Arc<Analyzed<F>>, degree: DegreeType) -> Arc<Analyzed
     }
 }
 
+mod sub_prover {
+    use std::{
+        collections::BTreeMap,
+        sync::{
+            mpsc::{Receiver, SyncSender},
+            Arc, Mutex,
+        },
+        thread::{Scope, ScopedJoinHandle},
+    };
+
+    use crate::{Backend, Error};
+    use powdr_executor::witgen::WitgenCallback;
+    use powdr_number::FieldElement;
+
+    /// Runs a prover until it either completes or challenges the caller, in which case the
+    pub fn run<'s, 'env, 'outer, F: FieldElement>(
+        scope: &'s Scope<'s, 'env>,
+        prover: &'env Mutex<Box<dyn Backend<'outer, F> + 'outer>>,
+        witness: Vec<(String, Vec<F>)>,
+    ) -> RunStatus<'s, F>
+    where
+        'outer: 'env,
+    {
+        let (challenge_sender, challenge_receiver) = std::sync::mpsc::sync_channel(0);
+        let (response_sender, response_receiver) = std::sync::mpsc::sync_channel(0);
+
+        let thread = scope.spawn(move || {
+            // We must ensure the backend can call the callback from
+            // multiple threads (needed by Plonky3, as it requires
+            // WitgenCallback to be Sync). So we wrap its Receiver in Mutex.
+            let response_receiver = Mutex::new(response_receiver);
+
+            let callback = WitgenCallback::<F>::new(Arc::new(move |_, challenge, _| {
+                // Lock first, to tie one challenge to one response.
+                let receiver_guard = response_receiver.lock().unwrap();
+                challenge_sender.send(challenge).unwrap();
+                receiver_guard.recv().unwrap()
+            }));
+
+            // TODO: this witness will be held in memory until the end of the
+            // proof, even if it's not needed anymore. We should probably change
+            // this API so the Vec is moved into the prover, and returned in the
+            // callback and result.
+            prover.lock().unwrap().prove(&witness, None, callback)
+        });
+
+        SubProver {
+            thread,
+            challenge_receiver,
+            response_sender,
+        }
+        .wait()
```

Comment:
Does calling `wait()` here mean that the `run` function only returns after the machine has committed to the first-phase witness? Maybe it would be nice to run all first-phase commitments in parallel?

Reply:
Yes. That was deliberate, as I assumed each prover would try to use all the processors/resources available while running. But we can change to run all in parallel, or at most N in parallel.

Reply:
How to do that? The proving is split in two halves. Report each half independently?

Reply:
Or add up the two halves?